dhall-json/src/Dhall/JSONToDhall.hs

```diff
@@ -369,13 +380,257 @@ typeCheckSchemaExpr compileException expr =
+    type should be
+
+    This is used by @{json,yaml}-to-dhall@ if the user does not supply a schema
+    on the command line
+-}
+inferSchema :: Value -> Schema
+inferSchema (A.Object m) =
+    let convertMap = Data.Map.fromList . HM.toList
+
+    in (Record . RecordSchema . convertMap) (fmap inferSchema m)
+inferSchema (A.Array xs) =
+    List (Foldable.fold (fmap inferSchema xs))
+inferSchema (A.String _) =
+    Text
+inferSchema (A.Number n) =
+    case floatingOrInteger n of
+        Left (_ :: Double) -> Double
+        Right (integer :: Integer)
+            | 0 <= integer -> Natural
+            | otherwise    -> Integer
+inferSchema (A.Bool _) = do
+    Bool
+inferSchema A.Null = do
+    Optional mempty
+
+-- | A record type that `inferSchema` can infer
+newtype RecordSchema =
+    RecordSchema { getRecordSchema :: Data.Map.Map Text Schema }
+
+instance Semigroup RecordSchema where
+    RecordSchema l <> RecordSchema r = RecordSchema m
+      where
+        -- The reason this is not @Just (Optional s)@ is to avoid creating a
+        -- double `Optional` wrapper when unifying a @null@ field with an
+        -- absent field.
+        onMissing _ s = Just (s <> Optional mempty)
+
+        m = Data.Map.Merge.merge
+                (Data.Map.Merge.mapMaybeMissing onMissing)
+                (Data.Map.Merge.mapMaybeMissing onMissing)
+                (Data.Map.Merge.zipWithMatched (\_ -> (<>)))
+                l
+                r
+
+recordSchemaToDhallType :: RecordSchema -> Expr s a
+recordSchemaToDhallType (RecordSchema m) =
+    D.Record (Map.fromList (Data.Map.toList (fmap schemaToDhallType m)))
+
+{-| `inferSchema` will never infer a union type with more than one numeric
+    alternative
+
+    Instead, the most general alternative type will be preferred, which this
+    type tracks
+-}
+data UnionNumber
+    = UnionAbsent
+    -- ^ The union type does not have a numeric alternative
+    | UnionNatural
+    -- ^ The union type has a @Natural@ alternative
+    | UnionInteger
+    -- ^ The union type has an @Integer@ alternative
+    | UnionDouble
+    -- ^ The union type has a @Double@ alternative
+    deriving (Bounded, Eq, Ord)
+
+-- | Unify two numeric alternative types by preferring the most general type
+instance Semigroup UnionNumber where
+    (<>) = max
+
+instance Monoid UnionNumber where
+    mempty = minBound
+
+unionNumberToAlternatives :: UnionNumber -> [ (Text, Maybe (Expr s a)) ]
+unionNumberToAlternatives UnionAbsent  = []
+unionNumberToAlternatives UnionNatural = [ ("Natural", Just D.Natural) ]
+unionNumberToAlternatives UnionInteger = [ ("Integer", Just D.Integer) ]
+unionNumberToAlternatives UnionDouble  = [ ("Double" , Just D.Double ) ]
+
+{-| A union type that `inferSchema` can infer
+
+    This type will have at most three alternatives:
+
+    * A @Bool@ alternative
+    * Either a @Natural@, @Integer@, or @Double@ alternative
+    * A @Text@ alternative
```

Comment:
It's actually intentional, not due to a lack of effort.  The rationale was that if a simple type and a record collided in the schema then the inference algorithm treats that as a strong heuristic indication that the user meant to use weakly-typed JSON there.  In those cases, inferring a convoluted union type would likely be counterproductive.

Reply:
OK. I lack the experience to judge what schema users would want in these cases.

Would your rationale be worth recording in a comment, just in case other readers have the same question?